contracts/mesh-provider/src/ibc.rs

```diff
@@ -213,33 +223,110 @@ pub fn ack_list_validators(
+        channel_id: channel_id,
+        data: to_binary(&packet)?,
+        timeout: build_timeout(&env),
+    };
+    Ok(IbcBasicResponse::new().add_message(msg))
 }
 
 pub fn fail_stake(
-    _deps: DepsMut,
+    deps: DepsMut,
     // _staker is the staker Addr, not used by consumer
-    _staker: String,
-    _validator: String,
-    _amount: Uint128,
+    staker: String,
+    validator: String,
+    amount: Uint128,
 ) -> Result<IbcBasicResponse, ContractError> {
-    // TODO: release the bonded stake, adjust numer
-    unimplemented!();
+    let staker = deps.api.addr_validate(&staker)?;
+
+    if amount.is_zero() {
+        return Err(ContractError::ZeroAmount);
+    }
+
+    let mut val = VALIDATORS
+        .may_load(deps.storage, &validator)?
+        .ok_or_else(|| ContractError::UnknownValidator(validator.clone()))?;
+    let mut stake = STAKED
+        .may_load(deps.storage, (&staker, &validator))?
+        .unwrap_or_default();
+    stake.unstake_validator(&mut val, amount)?;
+    STAKED.save(deps.storage, (&staker, &validator), &stake)?;
+    VALIDATORS.save(deps.storage, &validator, &val)?;
+    Ok(IbcBasicResponse::new())
+}
+
+pub fn ack_unstake(
+    deps: DepsMut,
+    env: Env,
+    staker: String,
+    _validator: String,
+    amount: Uint128
+) -> Result<IbcBasicResponse , ContractError> {
+    let staker = deps.api.addr_validate(&staker)?;
+
+    if amount.is_zero() {
+        return Err(ContractError::ZeroAmount);
+    }
+
+    // create a future claim on number of shares (so we can adjust for later slashing)
+    let cfg = CONFIG.load(deps.storage)?;
+    let ready = env.block.time.plus_seconds(cfg.unbonding_period);
+    CLAIMS.create_claim(
+        deps.storage,
+        &staker,
+        amount,
+        Expiration::AtTime(ready),
+    )?;
+
+    // TODO: How should the user be notified of the failure?
+    Ok(IbcBasicResponse::new())
 }
 
 pub fn fail_unstake(
-    _deps: DepsMut,
+    deps: DepsMut,
     // _staker is the staker Addr, not used by consumer
-    _staker: String,
-    _validator: String,
-    _amount: Uint128,
+    staker: String,
+    validator: String,
+    amount: Uint128,
 ) -> Result<IbcBasicResponse, ContractError> {
-    // TODO: unrelease the bonded stake, remove claim
-    // Maybe we only make Claim on ack?
-    unimplemented!();
+    let staker = deps.api.addr_validate(&staker)?;
+
+    if amount.is_zero() {
+        return Err(ContractError::ZeroAmount);
+    }
+
+    // revert state changes associated with stake
+    let mut val = VALIDATORS
+        .may_load(deps.storage, &validator)?
+        .ok_or_else(|| ContractError::UnknownValidator(validator.clone()))?;
+    let mut stake = STAKED.load(deps.storage, (&staker, &validator))?;
+    stake.stake_validator(&mut val, amount);
+    STAKED.save(deps.storage, (&staker, &validator), &stake)?;
+    VALIDATORS.save(deps.storage, &validator, &val)?;
+
+    // TODO: How should the user be notified of the failure?
```

Comment:
IMO, we just add the relevant events / attributes, and leave it to the frontend.